Add build_requires support for conanfile.txt loader

tonka3000 opened the pull request:
This PR will add build_requires support for the conanfile.txt loader. This can be usefull in f.e. conan-cmake project [#67](https://github.com/conan-io/cmake-conan/issues/67).

Greetings  
Tonka

lasote commented:
Thanks for the contribution, we have to add some tests but we could help.

tonka3000 commented:
@lasote , i think i can close this PR, because of the discussion in [#64](https://github.com/conan-io/cmake-conan/issues/67) ?

memsharded commented:
We have discussed and it could make sense to add it, so no need to close it. Even if the final result could be the same, it is good to have them defined. It is important to note that this [build_rquires] section is just for the consuming project, not for the dependencies. For dependencies, the declaration should be in the profile.

To merge this it is needed:
- Add some tests
- Add the relevant docs.

Our backlog is full now, I am temptatively asigning it to 1.2, if you could contribute those missing pieces it would help, if not we might try to contribute them ourselves, depending on the backlog for 1.2. 

Many thanks!

tonka3000 commented:
Added section to the docs.

I've also added one the (conanfile.txt reader).

I wasn't able to add a test to the install tests, because i get many errors locally without adding any new test, do you know why?

memsharded commented on conans/model/conan_file.py:
```diff
@@ -50,6 +50,22 @@ def create_requirements(conanfile):
         raise ConanException("Error while initializing requirements. %s" % str(e))
 
 
+def create_build_requirements(conanfile):
```
I think you don't need to change conan_file.py at all. 
This PR should affect only to the text version of the conanfile. Conanfile.py already manages build_requires, but in a different place.

tonka3000 replied:
I will remove the changes from these file

memsharded commented on conans/client/loader_parse.py:
```diff
@@ -95,7 +95,7 @@ class ConanFileTextLoader(object):
         # Prefer composition over inheritance, the __getattr__ was breaking things
         self._config_parser = ConfigParser(input_text,  ["requires", "generators", "options",
-                                                         "imports"], parse_lines=True)
+                                                         "imports", "build_requires"], parse_lines=True)
```
This might be considered breaking in the sense that conanfile.txt that are created with this feature won't load at all in previous conan versions. But as it is a consumer file, and we can consider only forward compatibility for it, I think we are good. But I'd like to check with @lasote anyway.

lasote replied:
I think we are good in this case.

memsharded commented:
@tonka3000 You can have a look to the tests failing in CI-jenkins, I think most of them failing are due to the modifications to ``conan_file.py``. 

If you need help locally running the tests, please open a dedicated issue reporting what is failing while running the tests. Thanks!

tonka3000 commented:
@memsharded i've changed my gitbranch to develop and then back to my feature-branch, now the tests work as expected 😕 .

I've revert all changes from `conan_file.py` and move all to `ConanFileLoader`

memsharded commented:
Nice, tests pass now, thank you!

Still, this feature deserves an integration test. I will contribute one to your PR now.

memsharded commented:
I think now it looks good and relatively low risk, so maybe we can move it forward to 1.1 instead of 1.2?

Wdyt @lasote?

lasote commented:
Yes, but, do we have docs for this? It requires (pun LoL) to explain the difference between using requires and build requires in the consumer project (if exist) or at least conceptually what are they meaning or for what it should be used.

memsharded commented:
In the reference itself is a good place? Just expanding the proposed section?

tonka3000 commented:
@memsharded should i extend the docs section?

memsharded commented:
Yes, please, if you could do it, that would be great. Otherwise I will do it.
Just a couple of paragraphs could be enough, the key ideas are:
- In practice both requires and build_requires will be installed. conanfile.txt are intended for users building their projects, so build_requires will be necessary.
- But it is good and clean to express the intention of different types of dependencies, if the recipe is to be converted to conanfile.py, for example.
Thanks!
